### PR TITLE
add an aws chalice integration

### LINKED ISSUE #845: add support for aws chalice
AWS Chalice is a python micro-services framework, which hosts endpoints using aws api gateway and aws lambdas.

https://github.com/aws/chalice

A class can be added to route requests and return appropriate responses in a similar manner to the strawberry flask framework.

### PR DESCRIPTION
**Pull request by mcsheehan**
This adds an aws chalice integration with unit tests.
A server can be launched from the tests directory by making a chalice config and then running chalice local. The unit tests capture the usage of chalice.

Closes #845

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

File: strawberry/chalice/graphiql.py
Comment (BryceBeagle) [inline strawberry/chalice/graphiql.py:21]: Can we switch this to using modern `pathlib`?

File: strawberry/chalice/views.py
Comment (BryceBeagle) [inline strawberry/chalice/views.py:14]:
```suggestion
    def __init__(self, schema: BaseSchema, render_graphiql: bool = True):
        self._schema = schema
        self.graphiql = render_graphiql
```

File: strawberry/chalice/views.py
Comment (BryceBeagle) [inline strawberry/chalice/views.py:23]:
```suggestion
```

File: strawberry/chalice/views.py
Comment (BryceBeagle) [inline strawberry/chalice/views.py:41]: Will you get a `KeyError` with `.get()`?

File: strawberry/chalice/views.py
Comment (BryceBeagle) [inline strawberry/chalice/views.py:67]:
```suggestion
        return method == "POST"
```

Comment (BryceBeagle) [reply strawberry/chalice/views.py:67]: Not sure if we need dedicated functions for these.

File: strawberry/chalice/views.py
Comment (BryceBeagle) [inline strawberry/chalice/views.py:82]:
```suggestion
        return method == "GET"
```

File: strawberry/chalice/views.py
Comment (BryceBeagle) [inline strawberry/chalice/views.py:141]: When would these raise `TypeError`s?

File: strawberry/chalice/views.py
Comment (BryceBeagle) [inline strawberry/chalice/views.py:120]:
```suggestion
            if (
                self.has_html_been_asked_for(request.headers)
                and self.method_is_get(request.method)
            ):
```

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: strawberry/chalice/views.py
Comment (BryceBeagle) [reply strawberry/chalice/views.py:41]: or a `TypeError` for that matter

File: strawberry/chalice/graphiql.py
Comment (mcsheehan) [reply strawberry/chalice/graphiql.py:21]: Happy to. I was keeping it in the same style as the flask plugin (copy paste job) I'll change it now.

Comment (BryceBeagle) [reply strawberry/chalice/graphiql.py:21]: Yeah everyone seems to be copying that flask plugin, and every review I have to ask for many of the same changes ahaha. I should stop being lazy and go fix the flask plugin too

File: strawberry/chalice/views.py
Comment (mcsheehan) [reply strawberry/chalice/views.py:142]: If people put weird stuff into their json requests. It can happen. I.E. saying it's application/json then putting a string in as the request (not json)

File: strawberry/chalice/views.py
Comment (mcsheehan) [reply strawberry/chalice/views.py:41]: You're right no key error - type error maybe people can do weird stuff when they make requests.

Comment (mcsheehan) [reply strawberry/chalice/views.py]: I've removed the keyerror, but left the typeerror for now

File: strawberry/chalice/views.py
Comment (DoctorJohn) [reply strawberry/chalice/views.py:136]: Im pretty sure these `get`s won't raise TypeErrors, but `request.json_body` a few lines above could raise an error when parsing the raw body. (See Chalice source code: https://github.com/aws/chalice/blob/8df8b59aa759292ec065dae54a074731b11fa28e/chalice/app.py#L437)

Comment (mcsheehan) [reply strawberry/chalice/views.py]: Have made the above changes

File: strawberry/chalice/views.py
Comment (mcsheehan) [reply strawberry/chalice/views.py]: I have removed them

File: strawberry/chalice/views.py
Comment (mcsheehan) [reply strawberry/chalice/views.py:115]: Added the catch around the request.json_body. The TypeError will get raised if the payload is reported as application/json, but a string is provided I believe, so I've got a test case for that and left it in

File: strawberry/chalice/views.py
Comment (mcsheehan) [reply strawberry/chalice/views.py]: I'm trying to work out if it's possible to pass headers in (not as a caseinsensitivemapping) it certainly is for the application json content, so I've left the TypeError in to be safe.

File: docs/integrations/chalice.md
Comment (BryceBeagle) [inline docs/integrations/chalice.md:58]:
```suggestion
GraphiQL will be hosted on localhost:8000/graphql
```

File: docs/integrations/chalice.md
Comment (BryceBeagle) [inline docs/integrations/chalice.md:53]:
```suggestion
And then run `chalice local` to start the localhost
```

File: docs/integrations/chalice.md
Comment (BryceBeagle) [inline docs/integrations/chalice.md:47]: Type annotate the return type here (might just be `object`?)

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: docs/integrations/chalice.md
Comment (mcsheehan) [reply docs/integrations/chalice.md]: Done

**Review by BryceBeagle — APPROVED** [review]

**Comment by BryceBeagle** [comment]
This may require a few changes if #1066 is merged first

File: strawberry/chalice/graphiql.py
Comment (patrick91) [inline strawberry/chalice/graphiql.py:5]: that's pretty neat!

File: strawberry/chalice/views.py
Comment (patrick91) [inline strawberry/chalice/views.py:17]: do you need this if we already have a cache for render_graphiql_page?

**Review by patrick91 — APPROVED** [review]
Sorry it took me a while to review this! Looks great, I've only added a minor comment, but I think it is ready to merge ^^

File: strawberry/chalice/views.py
Comment (mcsheehan) [reply strawberry/chalice/views.py]: You're right the other cache covers this - I have now removed it.

File: strawberry/chalice/graphiql.py
Comment (mcsheehan) [reply strawberry/chalice/graphiql.py:5]: Thanks :D

File: RELEASE.md
Comment (patrick91) [inline RELEASE.md:1]:
```suggestion
Release type: minor
```

;)

**Comment by patrick91** [comment]
I'll merge this after I fix the mypy issue :D

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**
